components/kyma-environment-broker/internal/kubeconfig/handler.go

```diff
@@ -0,0 +1,131 @@
+package kubeconfig
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/httputil"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/storage"
+
+	"github.com/google/uuid"
+	"github.com/gorilla/mux"
+	"github.com/pivotal-cf/brokerapi/v8/domain"
+	"github.com/sirupsen/logrus"
+)
+
+const attachmentName = "kubeconfig.yaml"
+
+//go:generate mockery -name=KcBuilder -output=automock -outpkg=automock -case=underscore
+
+type KcBuilder interface {
+	Build(*internal.Instance) (string, error)
+}
+
+type Handler struct {
+	kubeconfigBuilder KcBuilder
+	instanceStorage   storage.Instances
+	operationStorage  storage.Operations
+	log               logrus.FieldLogger
+}
+
+func NewHandler(storage storage.BrokerStorage, b KcBuilder, log logrus.FieldLogger) *Handler {
+	return &Handler{
+		instanceStorage:   storage.Instances(),
+		operationStorage:  storage.Operations(),
+		kubeconfigBuilder: b,
+		log:               log,
+	}
+}
+
+func (h *Handler) AttachRoutes(router *mux.Router) {
+	router.HandleFunc("/kubeconfig/{operation_id}/instance/{instance_id}", h.GetKubeconfig).Methods(http.MethodGet)
+}
+
+type InfoResponse struct {
+	Message string
+}
+
+type ErrorResponse struct {
+	Error string
+}
+
+func (h *Handler) GetKubeconfig(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	instanceID := vars["instance_id"]
+	operationID := vars["operation_id"]
+
+	if !h.isValidUUID(operationID) {
+		h.handleErrorResponse(w, http.StatusBadRequest, errors.New("wrong operationID"))
+		return
+	}
+
+	if !h.isValidUUID(instanceID) {
+		h.handleErrorResponse(w, http.StatusBadRequest, errors.New("wrong instanceID"))
+		return
+	}
+
+	instance, err := h.instanceStorage.GetByID(instanceID)
+	if err != nil {
+		h.handleErrorResponse(w, http.StatusNotFound, err)
+		return
+	}
+
+	if instance.RuntimeID == "" {
+		h.handleInfoResponse(w, fmt.Sprintf("kubeconfig for instance %s does not exist. Provisioning could be in progress, please try again later.", instanceID))
+		return
+	}
+
+	operation, err := h.operationStorage.GetProvisioningOperationByID(operationID)
+	if err != nil {
+		h.handleErrorResponse(w, http.StatusNotFound, err)
+		return
+	}
+
+	if operation.InstanceID != instanceID {
+		h.handleErrorResponse(w, http.StatusBadRequest, errors.New("mismatch between operation and instance"))
+		return
+	}
+
+	switch operation.State {
+	case domain.InProgress:
```

Comment:
what about pending state?

Reply:
IMO `Pending` is a status of orchestration which is part of the update operation, during the update cluster already exist so kubeconfig could be downloaded. There is more [update statuses](https://github.com/kyma-project/control-plane/blob/main/components/kyma-environment-broker/common/orchestration/dto.go#L29) which I'm not supported. BTW this is a little mess to have statuses in two different places (domain from pivotal and orchestration from KEB)

Reply:
We have introduced a Pending state which indicates that the operation didn't really started. The GetProvisioningOperationByInstanceID returns the latest provisioning operation which could be triggered by unsuspension. Such operation has state Pending until every other operation has finished.
The Pending state does not match OSB API and we are not able to define it in the same place as others states.